### ISSUE TITLE
Freezing layers not working with TF optimizer

### ISSUE THREAD

**Issue by Overdrivr** [keras-team/keras#10222]
It seems that freezing layers is not working when optimization is performed through a TF optimizer.
Most likely related to #4471, #6646 and maybe #4674

Code for reproducing the issue:

```python
from keras.layers import *
from keras.models import Model
import tensorflow as tf
import numpy as np
from keras.backend import clear_session

def dataset_fn():
    feats = np.random.randint(0, high=10, size=(1, 9)).astype(np.float32)
    labels = np.random.randint(0, 10, size=(1, 1)).astype(np.float32)

    dataset = tf.data.Dataset.from_tensor_slices((feats, labels))
    #dataset = dataset.shuffle(100)
    dataset = dataset.repeat(10000)
    dataset = dataset.batch(2)

    return dataset

def print_weights(tensor, layer):
    return tensor

def model_fn(features, labels, mode, params):
    x = Input(shape=(9,))
    layer = Dense(32)
    y = layer(x)

    model = Model(x, y)

    for l in model.layers:
        l.trainable = False

    model.compile(optimizer='rmsprop', loss='mse')

    if mode == tf.estimator.ModeKeys.TRAIN:
        optimizer = tf.train.AdamOptimizer(learning_rate=1e-4)

        res = model(features)
        res = tf.Print(res, [res], 'Output response:', summarize=6)
        loss = tf.reduce_sum(res)
        loss = tf.Print(loss, [loss], 'Loss:')

        return tf.estimator.EstimatorSpec(
            mode=tf.estimator.ModeKeys.TRAIN,
            loss=loss,
            train_op=optimizer.minimize(loss, tf.train.get_or_create_global_step()))

estimator = tf.estimator.Estimator(model_dir="./model", model_fn=model_fn)

for i in range(10):
    estimator.train(input_fn=dataset_fn, steps=10)
    clear_session()
```

Layers are frozen, and model is compiled, yet the weights change during training while I expected them not to:

```
Output response:[[-176.963181 -178.108612 -179.35141 -175.689926 -180.10643 -166.333939]...]
Loss:[-11203.0391]
Output response:[[-176.967072 -178.112488 -179.355301 -175.693817 -180.110306 -166.33783]...]
Loss:[-11203.2871]
Output response:[[-176.971039 -178.116455 -179.359253 -175.697784 -180.114288 -166.341797]...]
Loss:[-11203.541]
Output response:[[-176.975098 -178.120499 -179.363327 -175.701843 -180.118332 -166.34584]...]
Loss:[-11203.8008]
Output response:[[-176.979248 -178.124664 -179.367462 -175.705963 -180.122482 -166.349991]...]
Loss:[-11204.0664]
Output response:[[-176.983459 -178.12883 -179.371643 -175.710175 -180.126694 -166.354172]...]
Loss:[-11204.335]
Output response:[[-176.987686 -178.133102 -179.375931 -175.714447 -180.130951 -166.358444]...]
Loss:[-11204.6074]
Output response:[[-176.99202 -178.137421 -179.380219 -175.71875 -180.135254 -166.362762]...]
Loss:[-11204.8838]
Output response:[[-176.996399 -178.141815 -179.384613 -175.723129 -180.139633 -166.367142]...]
```
The output of the model is evolving, and the loss as well. So if I'm right, it does mean the weights of the model are changing.

Any idea of what is going on ? I'm facing this issue in a GAN-like architecture, with two networks where I want only one of them to be trained. Any workaround ?

I don't know how many people mix TF and Keras like this, but this got me stuck for a couple of days, before realizing both nets were actually being trained.

If this behavior is expected, it might be worth mentioning it in the FAQ.

EDIT: Fixed bug in demo code
